e2etest/pkg/config/update.go

```diff
@@ -0,0 +1,295 @@
+// SPDX-License-Identifier:Apache-2.0
+
+package config
+
+import (
+	"context"
+	"time"
+
+	operatorv1beta1 "github.com/metallb/metallb-operator/api/v1beta1"
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v2"
+	v1 "k8s.io/api/core/v1"
+	apierr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type Updater interface {
+	Update(f File) error
+}
+
+type Cleaner interface {
+	Clean() error
+}
+
+type UpdateCleaner interface {
+	Updater
+	Cleaner
+}
+
+type cmUpdateCleaner struct {
+	clientset.Interface
+	name      string
+	namespace string
+}
+
+func NewConfigMap(c clientset.Interface, ns string) *cmUpdateCleaner {
+	return &cmUpdateCleaner{
+		Interface: c,
+		name:      "config",
+		namespace: ns,
+	}
+}
+
+func (c cmUpdateCleaner) Update(f File) error {
+	resData, err := yaml.Marshal(f)
+	if err != nil {
+		return errors.Wrapf(err, "Failed to marshal MetalLB ConfigMap data")
+	}
+
+	_, err = c.CoreV1().ConfigMaps(c.namespace).Update(context.TODO(), &v1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      c.name,
+			Namespace: c.namespace,
+		},
+		Data: map[string]string{"config": string(resData)},
+	}, metav1.UpdateOptions{})
+
+	if err != nil {
+		return errors.Wrapf(err, "Failed to update MetalLB ConfigMap")
+	}
+
+	return nil
+}
+
+func (c cmUpdateCleaner) Clean() error {
+	return c.Update(File{})
+}
+
+type operatorUpdateCleaner struct {
+	client.Client
+	namespace string
+}
+
+func NewOperator(r *rest.Config, ns string) (*operatorUpdateCleaner, error) {
+	myScheme := runtime.NewScheme()
+
+	if err := operatorv1beta1.AddToScheme(myScheme); err != nil {
+		return nil, err
+	}
+
+	cl, err := client.New(r, client.Options{
+		Scheme: myScheme,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &operatorUpdateCleaner{
+		Client:    cl,
+		namespace: ns,
+	}, nil
+}
+
+func (o operatorUpdateCleaner) Update(f File) error {
+	for _, a := range f.Pools {
+		err := o.createPool(a)
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, bp := range f.BFDProfiles {
+		err := o.createBFDProfile(bp)
+		if err != nil {
+			return err
+		}
+	}
+
+	/*
+		Since a peer doesn't have a name we need to clean them before creating.
+		Without this the webhook fails the request because we are trying to create
+		a BGPPeer with the same configuration under a different (generated) name.
+		TODO: is there a better way to handle this?
+	*/
+	err := o.DeleteAllOf(context.Background(), &operatorv1beta1.BGPPeer{}, client.InNamespace(o.namespace))
```

Comment:
When does this happen? Does it mean that we are leaking peers between tests?
I assumed each test cleans the config afterwards.

Reply:
It happens when running the `should work after subsequent configuration updates` tests.
The tests create configuration with the same peers but adding an addresspool each iteration. 
The current implementation doesn't care about this since it always writes over the current ConfigMap, 
but with CRs this causes a conflict because it also recreates the BGPPeers (with the same config) in each iteration.